File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
n, k = map(int, input().split())
infos = [list(map(int, input().split())) for _ in range(n)]

def solution(n, k, infos):

    # 냅색
    # 행 : 가방(item - 인덱스)
    # 열 : 무게
    # 아이템 : 가치
    # dp[i][j] = max(dp[i][j], dp[i][j - w] + v) 
    # dp[i - 1][j] 가 맞지 않아? 그 전 행을 그대로 복사하기 때문에 dp[i - 1][j] = dp[i][j] 동일한 값을 가지게 된다.
    inf = int(1e18)
    # k 무게 초과는 볼 필요가 없음 / 행 : 첫번째 가방 선택~
    dp = [[-inf for _ in range(k + 1)] for _ in range(n + 1)]
    # 1. 기초셋팅
    dp[0][0] = 0
    # dp에서 행을 n + 1로 했음
    # 0base -> 1base
    infos = [[0, 0]] + infos
    
    for i in range(1, n + 1): 
        w, v = infos[i] # 0base -> 1base
        for j in range(k + 1):
            dp[i][j] = dp[i - 1][j] #1base
            if j - w >= 0:
                dp[i][j] = max(dp[i][j], dp[i - 1][j - w] + v)
    
    answer = -inf
    
    for i in range(1, n + 1):
        answer = max(answer, max(dp[i]))
        
    return answer

# 결론 : dp[i][j] = max(dp[i][j], dp[i - 1][j - w] + v) 
print(solution(n, k, infos))

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.*;
import java.util.*;

class Solution {
    public long solution(int n, int k, int[][] infos) {
        // 냅색
        // item : 가방
        // 행 : i번째 가방
        // 열 : 무게(제한있는거고)
        // dp에 들어가는 값 : 가치
        // dp[i][j] = i번째 가방을 고려햇을 때 무게 j일때 최고
        long[][] dp = new long[n + 1][k + 1];
        long inf = (long) 1e18;

        for(int i = 0; i < n + 1; i++) {
            for (int j = 0; j < k + 1; j++) {
                dp[i][j] = -inf;
            }
        }

        dp[0][0] = 0;

        for (int i = 1; i < n + 1; i++) {
            int w = infos[i - 1][0];
            int v = infos[i - 1][1];
            for (int j = 0; j < k + 1; j++) {
                // i번째 가방 고려...그 전 선택한 가방에서 j무게일 떼 최고가치
                dp[i][j] = dp[i - 1][j];
                // i 선택 + w만큼 채우기..
                if (j - w >= 0) {
                    dp[i][j] = Math.max(dp[i][j], dp[i - 1][j - w] + v);
                }
            }
        }

        long answer = -inf;

        for (int j = 0; j < k + 1; j++) {
            answer = Math.max(answer, dp[n][j]);
        }

        return answer;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());  // 물품의 수
        int k = Integer.parseInt(st.nextToken());  // 배낭의 최대 무게

        int[][] infos = new int[n][2];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            infos[i][0] = Integer.parseInt(st.nextToken());  // 무게
            infos[i][1] = Integer.parseInt(st.nextToken());  // 가치
        }

        Solution sol = new Solution();
        System.out.println(sol.solution(n, k, infos));

    }
}